docs/source/notes/modules.rst

```diff
@@ -0,0 +1,433 @@
+     (l0): MyLinear()
+     (l1): MyLinear()
+   ))
+   ('net.l0', MyLinear())
+   ('net.l1', MyLinear())
+
+Sometimes, it's necessary for a module to dynamically define submodules.
+The :class:`~torch.nn.ModuleList` and :class:`~torch.nn.ModuleDict` modules are useful here; they
+register submodules from a list or dict:
+
+.. code-block:: python
+
+   class DynamicNet(nn.Module):
+     def __init__(self, num_layers):
+       super().__init__()
+       self.linears = nn.ModuleList(
+         [MyLinear(4, 4) for _ in range(num_layers)])
+       self.activations = nn.ModuleDict({
+         'relu': nn.ReLU(),
+         'lrelu': nn.LeakyReLU()
+       })
+       self.final = MyLinear(4, 1)
+     def forward(self, x, act):
+       for linear in self.linears:
+         x = linear(x)
+       x = self.activations[act](x)
+       x = self.final(x)
+       return x
+
+   dynamic_net = DynamicNet(3)
+   sample_input = torch.randn(4)
+   output = dynamic_net(sample_input, 'relu')
+
+For any given module, its parameters consist of its direct parameters as well as the parameters of all submodules.
+This means that calls to :func:`~torch.nn.Module.parameters` and :func:`~torch.nn.Module.named_parameters` will
+recursively include child parameters, allowing for convenient optimization of all parameters within the network:
+
+.. code-block:: python
+
+   for parameter in dynamic_net.named_parameters():
+     print(parameter)
+   : ('linears.0.weight', Parameter containing:
+   tensor([[-1.2051,  0.7601,  1.1065,  0.1963],
+           [ 3.0592,  0.4354,  1.6598,  0.9828],
+           [-0.4446,  0.4628,  0.8774,  1.6848],
+           [-0.1222,  1.5458,  1.1729,  1.4647]], requires_grad=True))
+   ('linears.0.bias', Parameter containing:
+   tensor([ 1.5310,  1.0609, -2.0940,  1.1266], requires_grad=True))
+   ('linears.1.weight', Parameter containing:
+   tensor([[ 2.1113, -0.0623, -1.0806,  0.3508],
+           [-0.0550,  1.5317,  1.1064, -0.5562],
+           [-0.4028, -0.6942,  1.5793, -1.0140],
+           [-0.0329,  0.1160, -1.7183, -1.0434]], requires_grad=True))
+   ('linears.1.bias', Parameter containing:
+   tensor([ 0.0361, -0.9768, -0.3889,  1.1613], requires_grad=True))
+   ('linears.2.weight', Parameter containing:
+   tensor([[-2.6340, -0.3887, -0.9979,  0.0767],
+           [-0.3526,  0.8756, -1.5847, -0.6016],
+           [-0.3269, -0.1608,  0.2897, -2.0829],
+           [ 2.6338,  0.9239,  0.6943, -1.5034]], requires_grad=True))
+   ('linears.2.bias', Parameter containing:
+   tensor([ 1.0268,  0.4489, -0.9403,  0.1571], requires_grad=True))
+   ('final.weight', Parameter containing:
+   tensor([[ 0.2509], [-0.5052], [ 0.3088], [-1.4951]], requires_grad=True))
+   ('final.bias', Parameter containing:
+   tensor([0.3381], requires_grad=True))
+
+It's also easy to move all parameters to a different device or change their precision using
+:func:`~torch.nn.Module.to`:
+
+.. code-block:: python
+
+   dynamic_net.to(device='cuda', dtype=torch.float64)
```

Comment:
nit: I think it's really rare to want to do this, for a simple tutorial it's probably better to just do this in two separate steps which will more closely resemble a single call that people do.